### PR TITLE
Introducing Oracle Database backed state store component

### PR DESCRIPTION
**Pull request by lucasjellema**
Signed-off-by: lucasjellema <lucasjellema@gmail.com>

# Description

This PR introduces a State Store component backed by Oracle Database (including OCI Autonomous Database service). The code is implemented following the example of the PostgreSQL state store component - with Oracle Database specific SQL and connection method & driver.

(this PR replaces https://github.com/dapr/components-contrib/pull/1448)

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: [#1444](https://github.com/dapr/components-contrib/issues/1444)

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs# [2109](https://github.com/dapr/docs/issues/2109_

### CODE REVIEW

**Comment by yaron2** [comment]
@daixiang0 can you please review this? we want to get this in for 1.7.

File: state/oracledatabase/oracledatabaseaccess.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	o.metadata = metadata
+	if val, ok := metadata.Properties[connectionStringKey]; ok && val != "" {
+		o.connectionString = val
+	} else {
+		o.logger.Error("Missing Oracle Database connection string")
+
+		return fmt.Errorf(errMissingConnectionString)
+	}
+	if val, ok := o.metadata.Properties[oracleWalletLocationKey]; ok && val != "" {
+		o.connectionString += "?TRACE FILE=trace.log&SSL=enable&SSL Verify=false&WALLET=" + url.QueryEscape(val)
```
Comment (daixiang0) [inline state/oracledatabase/oracledatabaseaccess.go:72]: Is it safe enough? Is there an existed lib to format SQL string?

File: state/oracledatabase/oracledatabaseaccess.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	db, err := sql.Open("oracle", o.connectionString)
+	if err != nil {
+		o.logger.Error(err)
+
+		return err
+	}
+
+	o.db = db
+
+	if pingErr := db.Ping(); pingErr != nil {
```
Comment (daixiang0) [inline state/oracledatabase/oracledatabaseaccess.go:83]: Now `Ping()` always return nil, how about comment this logic and add TODO?

File: state/oracledatabase/oracledatabaseaccess.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+		tx, err = o.db.Begin()
+		if err != nil {
+			return fmt.Errorf("failed to start database transaction : %w", err)
+		}
+	} else { // join the transaction passed in.
+		tx = o.tx
+	}
+	etag := uuid.New().String()
+	// QUESTION: only check for etag if FirstWrite specified - or always when etag is supplied??
+	if req.Options.Concurrency != state.FirstWrite {
```
Comment (daixiang0) [inline state/oracledatabase/oracledatabaseaccess.go:171]: /cc @artursouza

File: state/oracledatabase/oracledatabaseaccess.go (hunk 2 above)
Comment (lucasjellema) [reply state/oracledatabase/oracledatabaseaccess.go:87]: Well spotted, an omission on my part. I have now implemented the Ping method - also in the mock object - and added a unit test.

**Comment by lucasjellema** [comment]
I would like to request help. I do not know how to resolve the issues reported on several go.sum files.

I have ran the Development Container, applied the changes, ran linting and testing. Ran `make make modtidy-all` and ran `make modtidy-all check-diff`. Locally, this did not result in differences. I then committed all changed go,.sum files. And the check failed on `make modtidy-all check-diff`. 

Can anyone suggest how I can move forward? What steps to take?

**Comment by lucasjellema** [comment]
Failing check on pubsub.kafka certification  - I do not know how to resolve this one. Please someone help.

File: state/oracledatabase/oracledatabaseaccess.go (hunk 3 above)
Comment (lucasjellema) [reply state/oracledatabase/oracledatabaseaccess.go:175]: @daixiang0  @artursouza  this has been resolved as per Discord message: 

Only check for etag if FirstWrite specified - as per Discord message thread https://discord.com/channels/778680217417809931/901141713089863710/938520959562952735. 

I will update the code and extend the integration and unit test to this effect. Commit to follow shortly.

Comment (lucasjellema) [reply state/oracledatabase/oracledatabaseaccess.go:175]: committed updated code and integration test

**Review by daixiang0 — APPROVED** [review]
LGTM overall

/cc @yaron2 @artursouza

**Comment by yaron2** [comment]
Thanks @lucasjellema

**Pull request merged**